### PR TITLE
Support full-iteration CUDA graphs with MFSDP v2

### LINKED ISSUE #6152: Verify MFSDP v2 works with MCore's GraphableMegatronModule
After https://github.com/NVIDIA/Megatron-LM/pull/5865 is merged, we can/should test MCore's full-iteration CG features, e.g., those layers that inherit from
https://github.com/NVIDIA/Megatron-LM/blob/076f61fc7be6dd0d8a7a95f99cc66dd775ac5c42/megatron/core/transformer/module.py#L159

For now, I don't mind reviewing PRs with https://github.com/NVIDIA/Megatron-LM/pull/5865 as a diffbase.

### PR DESCRIPTION
**Pull request by wujingyue**
Closes #6152.

Adds full-iteration CUDA-graph support for Megatron-FSDP v2.

- Preserve MFSDP v2 optimizer gradient bindings across CUDA-graph replay.
- Add an 8-GPU HybridModel functional recipe with bit-exact determinism coverage. The dense Llama-3-shaped proxy intentionally excludes MoE, whose data-dependent expert-dispatch buffers are not capture-safe.
- Add a two-rank unit test that compares eager MFSDP v2 training with full-iteration and optimizer CUDA graphs enabled, including BF16 optimizer-state coverage.

### CODE REVIEW

File: megatron/core/optimizer/fully_sharded_optimizer.py [hunk 1]
```diff
@@ -204,16 +204,22 @@ def count_zeros(self) -> float:
...
+            getattr(model_chunk.config, "cuda_graph_impl", "none") != "none"
+            for model_chunk in self.model_chunks
+        ):
+            set_to_none = False
+
         if not self.is_stub_optimizer:
             self.optimizer.zero_grad(set_to_none=set_to_none)
 
-        # Empty local DTensor shards are filtered out of optimizer param groups
-        # as a TE FusedAdam workaround. A rank with no local optimizer params
```
Comment (wujingyue) [inline megatron/core/optimizer/fully_sharded_optimizer.py:212]: The fix for https://github.com/NVIDIA/TransformerEngine/issues/3207 has landed in Megatron-LM main, so I removed this.

File: tests/functional_tests/test_cases/gpt/gpt3_mfsdp_v2_cuda_graph/model_config.yaml [hunk 2]
```diff
@@ -0,0 +1,104 @@
+# Dense GPT on Megatron-FSDP v2 with full-iteration CUDA graphs.
```
Comment (wujingyue) [inline tests/functional_tests/test_cases/gpt/gpt3_mfsdp_v2_cuda_graph/model_config.yaml]: Say MFSDP v2

File: tests/functional_tests/test_cases/gpt/gpt3_mfsdp_v2_cuda_graph/model_config.yaml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+#
+# Covers the dense half of #6152 (verify MFSDP v2 works with MCore's
+# GraphableMegatronModule). MoE is intentionally excluded: expert dispatch sizes
+# buffers from data-dependent token counts via a host sync, which is illegal
+# during graph capture. That is a documented MoE limitation independent of FSDP
+# (see megatron/core/transformer/moe/README.md), so a dense model is what
+# isolates MFSDP v2's own all-gather/reduce-scatter mechanics.
```
Comment (wujingyue) [inline tests/functional_tests/test_cases/gpt/gpt3_mfsdp_v2_cuda_graph/model_config.yaml]:
```suggestion
# Covers #6152
```

**Review by guihong-nv — APPROVED** [review]

File: megatron/core/optimizer/fully_sharded_optimizer.py [hunk 4]
```diff
@@ -204,16 +204,22 @@ def count_zeros(self) -> float:
...
+        # install_sharded_grads() binds .grad to a persistent main_grad view from
+        # Python during backward, and graph replay re-executes only GPU kernels. So
+        # unbinding here is never undone on a replayed step: .grad stays None, the
+        # optimizer skips every parameter, and the run silently stops training with
+        # a 0.0 grad norm. Zero in place to keep the binding alive.
+        #
+        # Overriding rather than asserting: ChainedOptimizer forwards its own True
+        # default positionally, so set_to_none is always True here regardless of
+        # caller intent, and an assert would fire on every graphed step.
+        if any(model_chunk.config.cuda_graph_impl != "none" for model_chunk in self.model_chunks):
```
Comment (skyw) [inline megatron/core/optimizer/fully_sharded_optimizer.py:217]: Q: it really use "none" string instead of None?

File: tests/functional_tests/test_cases/hybrid/hybrid_llama3_proxy_mfsdp_v2_cuda_graph/golden_values_dev_dgx_h100.json [hunk 5]
```diff
@@ -0,0 +1,286 @@
+{
```
Comment (skyw) [inline tests/functional_tests/test_cases/hybrid/hybrid_llama3_proxy_mfsdp_v2_cuda_graph/golden_values_dev_dgx_h100.json:1]: Q: a training test doesn't seem to provide more protection than a unit test to simple change in this PR?

**Review by skyw — APPROVED** [review]

File: megatron/core/optimizer/fully_sharded_optimizer.py (hunk 4 above)
Comment (wujingyue) [reply megatron/core/optimizer/fully_sharded_optimizer.py:217]: Yep -- https://github.com/NVIDIA/Megatron-LM/blob/fce0676d6b5438337ceaefa999577b0c7b95fc13/megatron/core/transformer/transformer_config.py#L1124

Should have been an enum instead

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 6]
```diff
@@ -13,8 +13,12 @@
 
 import megatron.core.distributed.fsdp.mcore_fsdp_adapter as mcore_fsdp_adapter
 from megatron.core.distributed import DistributedDataParallelConfig
-from megatron.core.distributed.fsdp.mcore_fsdp_adapter import FullyShardedDataParallel
+from megatron.core.distributed.fsdp.mcore_fsdp_adapter import (
+    FullyShardedDataParallel,
+    FullyShardedDataParallelV2,
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: Needed?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 7]
```diff
@@ -69,7 +73,9 @@ class TestMcoreAdapterDense:
     def setup_method(self):
         Utils.initialize_model_parallel(1, 1)
         self.pg_collection = ProcessGroupCollection.use_mpu_process_groups()
-        model_parallel_cuda_manual_seed(1234)
+        # The CUDA-graph test may leave a TE tracker in process-global state; recreate the
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: This test doesn't do cuda graph any more. Still needed?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 8]
```diff
@@ -229,24 +235,20 @@ def test_build_train_and_step(self, optimizer_cuda_graph):
         reference_model = _build_block(config)
         model = _build_block(config)
         model.load_state_dict(reference_model.state_dict())
-        # get_megatron_optimizer expects every model chunk to expose ddp_config. The
-        # reference model remains unwrapped/unsharded, so it cannot use the
-        # DistributedOptimizer path that expects DDP/FSDP buffer metadata.
+        # The unwrapped reference cannot use the DistributedOptimizer path, which
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: Unwrapped?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 9]
```diff
@@ -290,23 +284,11 @@ def test_build_train_and_step(self, optimizer_cuda_graph):
...
-            reference_losses.append(torch.stack(microbatch_losses).mean())
-
-        losses = []
-        with torch.profiler.profile(activities=[torch.profiler.ProfilerActivity.CPU]) as prof:
+        def run(model, optimizer, is_mfsdp):
+            losses = []
             for microbatches in steps:
-                model.zero_grad_buffer()
+                if is_mfsdp:
+                    model.zero_grad_buffer()
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: Needed?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 10]
```diff
@@ -473,6 +444,149 @@ def global_norm(local_tensors) -> float:
...
+    def teardown_method(self):
+        # The wrappers store capture state globally. Reset it so the next test captures its
+        # own work instead of replaying this test's graph.
+        OptimizerCudaGraphWrapper.curr_iteration = 0
+        OptimizerCudaGraphWrapper.cuda_graph = None
+        OptimizerCudaGraphWrapper.result = None
+        FullCudaGraphWrapper.curr_iteration = {'training': 0, 'validation': 0}
+        FullCudaGraphWrapper.cuda_graph = {'training': None, 'validation': None}
+        FullCudaGraphWrapper.result = {'training': None, 'validation': None}
+        StaticBufferLoader.static_buffers = {'training': [], 'validation': []}
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py:457]: Should this be done in setup_method instead?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 11]
```diff
@@ -473,6 +444,149 @@ def global_norm(local_tensors) -> float:
...
+                optimizer.step = OptimizerCudaGraphWrapper(
+                    optimizer.step, cuda_graph_warmup_steps=1
+                )
+            return model, optimizer
+
+        def forward_backward(*, model, data_iterator, num_microbatches, **_):
+            microbatch_losses = []
+            for _ in range(num_microbatches):
+                batch = next(data_iterator[0])
+                output = model[0](hidden_states=batch["hidden_states"], attention_mask=None)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py:526]: [0]?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 12]
```diff
@@ -473,6 +444,149 @@ def global_norm(local_tensors) -> float:
...
+            assert isinstance(optimizer, FullyShardedOptimizer)
+            optimizer.reload_model_params()
+            if enable_cuda_graph:
+                # The factory makes FusedAdam capturable, while training installs this wrapper.
+                optimizer.step = OptimizerCudaGraphWrapper(
+                    optimizer.step, cuda_graph_warmup_steps=1
+                )
+            return model, optimizer
+
+        def forward_backward(*, model, data_iterator, num_microbatches, **_):
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: What's `**_` for?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 13]
```diff
@@ -473,6 +444,149 @@ def global_norm(local_tensors) -> float:
...
+            losses = []
+            for microbatches in steps:
+                model.zero_grad_buffer()
+                optimizer.zero_grad(set_to_none=True)
+                microbatch_losses = forward_backward(
+                    model=[model],
+                    data_iterator=[iter([{"hidden_states": batch} for batch in microbatches])],
+                    num_microbatches=len(microbatches),
+                    seq_length=None,
+                    forward_only=False,
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py:550]: Needed?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 14]
```diff
@@ -473,6 +444,149 @@ def global_norm(local_tensors) -> float:
...
+
+        def run(model, optimizer, forward_backward) -> torch.Tensor:
+            losses = []
+            for microbatches in steps:
+                model.zero_grad_buffer()
+                optimizer.zero_grad(set_to_none=True)
+                microbatch_losses = forward_backward(
+                    model=[model],
+                    data_iterator=[iter([{"hidden_states": batch} for batch in microbatches])],
+                    num_microbatches=len(microbatches),
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py:547]: Why? Can't simply exhaust data_iterator?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 15]
```diff
@@ -502,7 +616,9 @@ def setup_method(self):
             embd=None,
             pos_embd=None,
         )
-        model_parallel_cuda_manual_seed(1234)
+        # The preceding CUDA-graph class uses a TE tracker; recreate the default tracker for
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py]: ditto -- does this test exercise cuda graph at all?

File: tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py [hunk 16]
```diff
@@ -214,8 +215,8 @@ def record_fully_shard_context(*args, **kwargs):
 
         assert fully_shard_context_calls == [True]
 
-    @pytest.mark.parametrize("optimizer_cuda_graph", [False, True], ids=["eager", "cuda_graph"])
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_mcore_adapter.py:217]: optimizer_cuda_graph is covered by the unit test added.

**Review by Phlip79 — APPROVED** [review]
What about GB200 functional test?

**Review by yaoyu-33 — APPROVED** [review]

**Comment by wujingyue** [comment]
> What about GB200 functional test?

I don't think we need it before https://github.com/NVIDIA/Megatron-LM/issues/5615

**Review by jaredcasper — APPROVED** [review]

File: megatron/core/optimizer/fully_sharded_optimizer.py (hunk 1 above)
Comment (Autumn1998) [reply megatron/core/optimizer/fully_sharded_optimizer.py:212]: A naive question: don’t we need to perform some checks and assertions on the TE version?

Comment (wujingyue) [reply megatron/core/optimizer/fully_sharded_optimizer.py:212]: I don't know. I asked in https://github.com/NVIDIA/Megatron-LM/pull/7112#discussion_r3960712111 and nobody has responded yet.

**Review by balasaajay — APPROVED** [review]

File: megatron/core/optimizer/fully_sharded_optimizer.py (hunk 1 above)
Comment (wujingyue) [reply megatron/core/optimizer/fully_sharded_optimizer.py:212]: https://github.com/NVIDIA/Megatron-LM/pull/7216
